doc/subsystems/networking/net-app-api.rst

```diff
@@ -0,0 +1,199 @@
+.. _net_app_api:
+
+Network Application API
+#######################
+
+The Network Application API allows applications to
+
+* **Initialize** the application for networking use. This means for example
+  that if the application needs to have an IPv4 address, and if DHCPv4 is
+  enabled, then the net-app API will make sure that the device will get an
+  IPv4 address before the application is started.
+
+* **Set** various options for the networking subsystem. This means that if the
+  user has set options like IP addresses, IEEE 802.15.4 channel etc. in the
+  project configuration file, then those settings are applied to the system
+  before application starts.
+
+* **Create** a simple TCP/UDP server or client application. The net-app API
+  has functions that make it easy to create a simple TCP or UDP based network
+  application. The net-app API also provides transparent TLS and DTLS support
+  for the application.
+
+The net-app API functionality is enabled by `CONFIG_NET_APP` option.
+The current net-app API implementation is still experimental and may change and
+improve in future releases.
+
+Initialization
+##############
+
+The net-app API provides `net_app_init()` function that can configure the
+networking subsystem for the application. Following configuration options
+control how the configuration is done:
+
+* **CONFIG_NET_APP_AUTO_INIT** automatically configures the system according
+  other configuration options. User does not need to call `net_app_init()` in
+  this case as that function will be automatically called when the system
+  boots. This option is enabled by default.
+
+* **CONFIG_NET_APP_INIT_TIMEOUT** tells how long to wait for the network
+  configuration during the system boot. For example if DHCPv4 is enabled, and
+  if the IPv4 address discovery takes long or the DHCPv4 server is not found,
+  the system will resume booting after this amount of seconds
+
+* **CONFIG_NET_APP_NEED_IPV6** tells that the application needs IPv6
+  connectivity. The `net_app_init()` will wait until it is able to setup
+  IPv6 address for the system before continuing. This means that the IPv6
+  duplicate address detection (DAD) has finished and the system has properly
+  setup the IPv6 address.
+
+* **CONFIG_NET_APP_NEED_IPV6_ROUTER** tells that the application needs IPv6
+  router connectivity. The `net_app_init()` will wait until it receives router
```

Comment:
I'd add a layman explanation what "router connectivity" means. E.g. ", i.e. needs access to external networks (like the Internet)".